text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,305 @@
+const scope = new Scope();
+
+scope.getClient(); // <-- returns a noop client by default
+
+scope.setClient(client1);
+scope.getClient(); // <-- returns client1
+```
+
+The current scope may be kept similar to how we currently keep the current hub, but this is SDK specific and not part of this RFC.
+
+When calling `getScope()` before a scope was made the current one (=before init was called), we will return a scope for a noop client. 
+A noop client is a regular client that simply does not send anything.
+
+This way, the API for `getClient()` can always return a client, and users do not have to guard against this being undefined all the time. 
+We may also expose a util like `sentryIsInitialized()` that checks if the current client is a Noop client (which currently you could have checked as `getCurrentHub().getClient() === undefined`).
+
+If you want to have multiple isolated clients, you can achieve this easily with this new setup:
+
+```js
+const client1 = new Client();
+const client2 = new Client();
+
+const scope1 = new Scope();
+const scope2 = new Scope();
+
+scope1.setClient(client1);
+scope2.setClient(client2);
+
+scope1.captureException(); // <-- isolated from scope2
+```
+
+## Scopes
+
+Scopes behave similar to how they behave today. 
+When a scope is forked via `withScope()`, a new scope is created that inherits all data currently set on the parent scope.
+
+The main change to Scopes is that they do not push/pop anymore, but instead fork an execution context (in languages where this makes sense/is possible).
+Basically, `withScope()` should behave like `hub.run()` does today in languages that have execution context forking.
+
+`client.getScope()` should return the current scope of this client in the current execution context.
+
+From a users perspective, this should mostly not be noticeable - they can always run `getScope()` to get the current scope, or `withScope(callback)` to fork a new scope off the current scope.
+
+You can make a scope the current one via `setCurrentScope(scope)`, which should bind the scope to the current execution context (or a global, in SDKs without execution context). This is a replacement for the current APIs like `makeMain(hub)` or `setCurrentHub(hub)`.
+
+You can still clone scopes manually the same way as before, e.g. via `Scope.clone(oldScope)` or a similar API. In contrast to `withScope()`, this will _not_ fork an execution context.
+
+You can update the client of a scope via `scope.setClient(newClient)`. This will not affect any scope that has already been forked off this scope, but any scope forked off _after_ the client was updated will also receive the updated client.
+
+Every scope is always tied to a client.
+
+## Global Scope
+
+In addition to the currently active scope, there will also be a new special scope, the **Global Scope**.
+The global scope is _not_ the initial scope, but a special scope that belongs to a client and is applied to any event that belongs to this client.
```

Comment:
This would make the client stateful all of a sudden. Not the end of the world, but feels odd.

Reply:
Not necessarily. This does not mean that the global scope is defined on the client, but that the global scope has a reference to the client :) the mechanism to look up the global scope can/should probably be more in parallel to getting the active scope. Think of it like this:

```js
const executionContext1; // this is one execution context somewhere
const executionContext2; // this is some other execution context somewhere else

executionContext1.globalScope --> Scope
executionContext2.activeScope --> Scope
executionContext3.isolationScope --> Scope
```

So all three of these scopes can be put on the execution context (or wherever the active scope is put). How _exactly_ this is done would be up to the SDK/language, but the idea is that the client remains stateless.

Reply:
I was also initially confused by this, but I think it's just a matter of wording this a bit differently. It says:

> but a special scope that belongs to a client and is applied to any event that belongs to this client

This makes it sound like the client owns the scope, even though technically it's the other way around. Maybe we can write something like

```suggestion
The global scope is _not_ the initial scope, but a special scope that has a client and is applied to any event that belongs to its client.
```

Reply:
The tricky thing to convey (and actually, we need to specifically spec this out), is that each client has a separete global scope. 

At least, I think this is what makes most sense 🤔 When talking with @mitsuhiko both options are possible (global scope is _truly_ global, or global scope is basically a "client scope"). 

Personally I tend to the idea that each client should have it's own global scope - simply because thinking about e.g. MFE and similar, we probably don't want to "pollute" other clients when doing `getGlobalScope().setTag()`. 

Wording wise, we could say:

> The global scope is _not_ the initial scope, but a separate, special scope. The global scope is different for each client, and is applied to every event.

Or something like this?